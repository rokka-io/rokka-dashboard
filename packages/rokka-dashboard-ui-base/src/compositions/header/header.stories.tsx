import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { Header } from './header';

storiesOf('Compositions', module)
  .addDecorator(story => <div style={{ width: '100vw', height: '64px' }}>{story()}</div>)
  .addParameters({
    info: 'The MobileMenuButton is not visible on the desktop viewport, please resize storybook accordingly.'
  })
  .add('Header', () => (
    <Header username={text('username', 'rokka')} onLogout={action('logout')} onToggleMenu={action('toggle menu')} />
  ));
