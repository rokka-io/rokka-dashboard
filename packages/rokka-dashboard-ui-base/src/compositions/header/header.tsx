import React, { Component } from 'react';
import styled from 'styled-components';
import { HeaderNavigation } from '../../components';
import { HeaderLogo, MobileMenuButton } from '../../elements';
import { colors } from '../../identity/colors/colors';
import { spaces } from '../../identity/spaces/spaces';

interface HeaderProps {
  /** Displayed username */
  username: string;
  /** Toggle menu callback */
  onToggleMenu(active: boolean): void;
  /** Logout callback */
  onLogout(): void;
}
interface HeaderState {
  /** Menu active indicator */
  menuActive: boolean;
}

export class Header extends Component<HeaderProps, HeaderState> {
  public state = { menuActive: false };

  public render() {
    const { menuActive } = this.state;
    const { username, onLogout } = this.props;

    return (
      <StyledHeader>
        <MobileMenuButton onClick={this.onMobileMenuButtonClick} active={menuActive} />
        <HeaderLogo width="6em" />
        <HeaderNavigation username={username} onLogout={onLogout} />
      </StyledHeader>
    );
  }

  private onMobileMenuButtonClick = () => {
    const { menuActive } = this.state;
    const { onToggleMenu } = this.props;
    this.setState(
      {
        menuActive: !menuActive
      },
      () => onToggleMenu(this.state.menuActive)
    );
  };
}

const StyledHeader = styled.header`
  position: fixed;
  left: 0;
  top: 0;
  right: 0;
  padding: 0 ${spaces.medium};
  height: 64px;
  line-height: 64px;
  background: ${colors.brand.primary};
  z-index: 99;
`;
