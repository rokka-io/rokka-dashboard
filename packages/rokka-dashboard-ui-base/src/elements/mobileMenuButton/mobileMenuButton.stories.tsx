import { action } from '@storybook/addon-actions';
import { boolean } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { colors } from '../../identity/colors/colors';
import { MobileMenuButton } from './mobileMenuButton';

storiesOf('Elements / Header', module)
  .addParameters({
    info: 'The MobileMenuButton is not visible on the desktop viewport, please resize storybook accordingly.'
  })
  .addDecorator(story => <div style={{ background: colors.brand.dark }}>{story()}</div>)
  .add('MobileMenuButton', () => <MobileMenuButton onClick={action('onClick')} active={boolean('active', false)} />);
