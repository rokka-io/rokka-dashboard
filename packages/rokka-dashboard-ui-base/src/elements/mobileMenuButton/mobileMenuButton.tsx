import React from 'react';
import styled from 'styled-components';
import { colors } from '../../identity/colors/colors';
import { media } from '../../identity/media/media';
import { spaces } from '../../identity/spaces/spaces';
import { fonts } from '../../identity/typography';

interface MobileMenuButtonProps {
  /** Whether button is active */
  active?: boolean;
  /** Callback */
  onClick?(): void;
}

export function MobileMenuButton({ onClick, active = false }: MobileMenuButtonProps) {
  return <StyledMobileMenuButton active={active} type="button" onClick={onClick} />;
}

interface StyledMobileMenuButtonProps {
  active?: boolean;
}

// TODO: move to own element
const LinkButton = styled.button`
  background-color: transparent;
  border: none;
  cursor: pointer;
  font-family: ${fonts.Families.brand};
  font-size: ${fonts.Sizes.small};
  line-height: inherit;
  display: inline;
  margin: 0;
  padding: 0;
`;

const StyledMobileMenuButton = styled(LinkButton)<StyledMobileMenuButtonProps>`
  transition: background 0.4s ease;
  display: inline-block;
  height: ${spaces.large};
  width: ${spaces.large};
  border-radius: 4px;
  color: ${colors.tints.white};
  font-size: ${fonts.Sizes.large};
  margin-right: ${spaces.medium};
  vertical-align: middle;
  position: relative;

  ${media.desktop`
    display: none;
  `}

  ${({ active }) =>
    active &&
    `
      border: 1px solid ${colors.tints.white};
      background: rgba(0,0,0,.1);
    `}

  &:after {
    top: 50%;
    left: 50%;
    position: absolute;
    content: 'â˜°';
    transform: translate(-50%, -50%);
  }
`;
