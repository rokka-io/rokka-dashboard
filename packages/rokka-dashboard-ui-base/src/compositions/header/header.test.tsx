import { shallow } from 'enzyme';
import React from 'react';

import { MobileMenuButton } from '../../elements/mobileMenuButton/mobileMenuButton';
import { Header } from './header';

describe('Compositions', () => {
  describe('Header', () => {
    it('renders the header correctly.', () => {
      const onLogout = jest.fn();
      const onToggleMenu = jest.fn();
      const component = shallow(<Header username="rokka-test" onLogout={onLogout} onToggleMenu={onToggleMenu} />);
      const beforeState = component.state();
      expect(beforeState).toStrictEqual({ menuActive: false });

      component.find(MobileMenuButton).simulate('click');

      expect(component).toMatchSnapshot();
      expect(onToggleMenu).toHaveBeenCalledTimes(1);

      const afterState = component.state();
      expect(afterState).toStrictEqual({ menuActive: true });
    });
  });
});
