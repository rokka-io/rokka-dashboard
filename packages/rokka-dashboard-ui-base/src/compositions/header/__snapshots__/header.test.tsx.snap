// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compositions Header renders the header correctly. 1`] = `
<styled.header>
  <MobileMenuButton
    active={true}
    onClick={[Function]}
  />
  <HeaderLogo
    width="6em"
  />
  <HeaderNavigation
    onLogout={[MockFunction]}
    username="rokka-test"
  />
</styled.header>
`;
